### PR TITLE
updating canary longform inference script with timestamps option

### PR DESCRIPTION
**Pull request by krishnacpuvvada**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds support for timestamps in long-form inference script for AED models.

**Collection**: [ASR]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
python speech_to_text_aed_chunked_infer.py \
    model_path=null \
    pretrained_name="nvidia/canary-1b-flash" \
    audio_dir="<(optional) path to folder of audio files>" \
    dataset_manifest="<(optional) path to manifest>" \
    output_filename="<(optional) specify output filename>" \
    chunk_len_in_secs=10.0 \
    batch_size=16 \
    decoding.beam.beam_size=1 \
    compute_timestamps=True
```
results for tedlium-3 long-form inference using above script

canary-1b-flash:
chunk_len_in_secs=40.0, compute_timestamps=False : 4.40 (no PnC)
chunk_len_in_secs=10.0, compute_timestamps=True : 5.14 (no PnC)

canary-180m-flash:
chunk_len_in_secs=40.0, compute_timestamps=False : 4.37 (no PnC)
chunk_len_in_secs=10.0, compute_timestamps=True : 5.69 (no PnC)

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [X] New Feature
- [ ] Bugfix
- [X] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by KunalDhawan — APPROVED** [review]
Great work, thanks! Left minor comments, but they are non-blocking, LGTM.

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 1]
```diff
@@ -1608,28 +1609,44 @@ def get_all_audios(self):
...
+        
+        missing_keys = [k for k in expected_slots if k not in sample]
+        if missing_keys:
+            raise RuntimeError(
+                f"We found sample that is missing the following keys: {missing_keys}"
+                f"Please ensure that every utterance in the input manifests contains these keys. Sample: {sample}"
+            )
+        
+        # fill optional slots
+        for k, v in self.asr_model.prompt.get_default_slot_values().items():
```
Comment (pzelasko) [inline nemo/collections/asr/parts/utils/streaming_utils.py:1636]: Please test it with canary-1b - it will likely fail, because PromptFormatter has no `get_default_slot_values` method. 

Let's try to implement this without changing PromptFormatter's API. It's a deliberate choice not to assign default slot values to prompt formatter class; instead each trained model should define its own default values (they can be different with the same prompt format). If you need to hardcode some here to get this released quickly, I am OK with that.

**Review by pzelasko — DISMISSED** [review]

File: nemo/collections/common/prompts/canary2.py [hunk 2]
```diff
@@ -108,6 +108,20 @@ class Canary2PromptFormatter(PromptFormatter):
...
+    DEFAULT_SLOT_VALUES = {
+        "decodercontext": "",
+        "emotion": "<|emo:undefined|>",
+        "itn": "<|noitn|>",
+        "timestamp": "<|notimestamp|>",
+        "diarize": "<|nodiarize|>",
+        "pnc": "<|pnc|>",  # consistent with canary1
+    }
+
+    def get_default_slot_values(self) -> dict:
```
[the remark this answers is not in the document]
Comment (krishnacpuvvada) [reply nemo/collections/common/prompts/canary2.py]: reverted the change to prompt formatter class.

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 3]
```diff
@@ -1672,12 +1691,102 @@ def transcribe(
...
+
+    def _join_text(self, merged_hypothesis, hypotheses):
+        merged_hypothesis.text = " ".join([h.text for h in hypotheses])
+        return merged_hypothesis
+    
+    def _join_y_sequence(self, merged_hypothesis, hypotheses):
+        merged_hypothesis.y_sequence = torch.cat([h.y_sequence for h in hypotheses])
+        return merged_hypothesis
+    
+    def _join_timestamp(self, merged_hypothesis, hypotheses):
```
[the remark this answers is not in the document]
Comment (krishnacpuvvada) [reply nemo/collections/asr/parts/utils/streaming_utils.py:1742]: added CICD test.
